### ISSUE TITLE
Interactive prompts should be written to stdout rather than stderr

### ISSUE THREAD

**Issue by krader1961** [att/ast#1380]
While debugging a problem using an `expect` based unit test I noticed I couldn't redirect stderr without losing the prompt. That's because the prompt is written to stderr. This is problematic. Bash seems to have the same mistake but zsh does not. This means you can't do `exec 2>/tmp/stderr` in an interactive shell and still see a command prompt. This is problematic because debugging tools like the `DPRINTF()` macro I added long ago and mechanisms like ASAN write to stderr. You should be able to redirect stderr to a file to capture the output of such mechanisms while still having a usable interactive shell; i.e., a shell where the prompt appears on the stdout stream.

**Comment by dannyweldon** [comment]
Regarding #1384, will this just affect interactive mode?  What about $PS2, $PS3 (for `select`) and $PS4?
Also, `dash` also behaves like `bash` and current `ksh`.

**Comment by krader1961** [comment]
@dannyweldon All prompts are written by the same top-level function so this change handles all of them.

As for dash and bash behaving like the current ksh stable release I think the question is why? Is it meant to ensure an interactive shell only works if both stdin, stdout and stderr are attached to a tty? The ksh code does not do that (I can't speak to dash or bash). So why use stderr for interactivity rather than stdout?

**Comment by kusalananda** [comment]
The `zsh` shell only seems puts prompts on standard output in interactive shells. In non-interactive shells, prompts seems to be written to standard error. IMHO, it would be a mistake to _always_ send prompts to standard output (i.e. also for non-interactive shells) as that would make the interactive dialogue between the user and a script part of the script's output.

**Comment by krader1961** [comment]
@kusalananda  Yes, all of these behaviors are problematic. Consider a script containing `read 'x?WTF: '` that is run by `zsh script 2>/tmp/file`. The prompt is written to stderr. Even though stdout is attached to the tty and stderr is not attached to the tty. So the user doesn't see a prompt. They just see an empty line. There are a few more variations on this theme where the various popular posix shells behave differently.

For me it seems like making it easy to capture stderr output independent of interactive I/O is more important than capturing stdout output independent of interactive I/O. This is obviously a highly subjective issue. So it won't hurt my feelings if people tell me this change is a "bridge too far". That is, a change that while it might be an improvement isn't worth the potential change in backward compatible behavior.

**Comment by siteshwar** [comment]
This change breaks POSIX compatibility. See [description for PS1](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html):

"Each time an interactive shell is ready to read a command, the value of this
variable shall be subjected to parameter expansion and written to standard
error."

**Comment by spagoveanu** [comment]
`zsh` does NOT write its prompt to stdout. It's writing it to a fd obtained by opening `ttyname(0)`. Any of `zsh >/dev/null 2>&1` or `cat | zsh` will tell you that. Just like running `exec >/dev/null` inside an interactive `zsh` shell.

**Comment by dannyweldon** [comment]
Here is one way to get around the issue:
```ksh
PS1='$(print -n "$ " > /dev/tty)'
```
